### PR TITLE
[linux-port] Use cross-platform dynamic lib load

### PR DESCRIPTION
**Pull request by pow2clk**
LLVM includes a DynamicLibrary class that works on multiple platforms
Rather than ifdefing between dlopen for non-Windows and LoadLibraryW
on Windows, using this class allows us to abstract platform differences
while using the same code.

Changed DxilLibInitialize to attempt to initialize dxil.dll so the
smart mutex inside the dynamic library object uses the same IMalloc
as the other ManagedStatic objects so they can all be freed in order.

### CODE REVIEW

**Comment by pow2clk** [comment]
I'm not sure I've seen appveyor fail arbitrarily, but I don't know what to make of this failure. The build succeeds, the tests succeeded on my Windows machine, but appveyor fails with error 127 right where it should run hcttest. 127 seems to indicate command not found. I didn't delete it. I don't know why it wouldn't be there. Regardless of the state of the code or presence or not of any other binaries, hcttest should have produced some kind of output.

**Comment by pow2clk** [comment]
Well twice is very unlikely to be a fluke. I've tried running the powershell script that is run just before the failure. Presumably that's what the failure is. It runs cleanly on my own machine. I've tried poking various holes into my own installation to try to get that return error, but nothing I rename or delete seems to cause exactly that problem. I really wish the script produced more output, but even the output the powershell script produces isn't included in the saved output.

**Comment by ehsannas** [comment]
At a first glance, it seems like a reasonable change to me. But I haven't used `llvm::sys::DynamicLibrary`, so I can't tell why the test failures are happening... I'll read into DynamicLibrary a bit more

File: tools/clang/tools/dxcompiler/dxillib.cpp [hunk 1]
```diff
@@ -25,15 +25,17 @@ static llvm::sys::Mutex *cs = nullptr;
...
+  cs->lock();
+  g_DllLibResult = g_DllSupport.InitializeForDll(L"dxil.dll", "DxcCreateInstance");
+  cs->unlock();
+#endif
   return S_OK;
 }
 
 HRESULT DxilLibCleanup(DxilLibCleanUpType type) {
   HRESULT hr = S_OK;
-  if (type == DxilLibCleanUpType::ProcessTermination) {
```
Comment (ehsannas) [inline tools/clang/tools/dxcompiler/dxillib.cpp]: Why did you remove this case?
shouldn't something like `m_dll = DynamicLibrary();` happen in this case too?

Comment (pow2clk) [reply tools/clang/tools/dxcompiler/dxillib.cpp]: It could, but it doesn't free any resources. The open Dynamic libraries persist until they are all closed by llvm shutdown. I suppose I could do that too keep m_dll from being used again. 

Is there a way to make the powershell script more verbose? It prints a lot when I run it locally, but it prints almost nothing in the automatic appveyor output. I'm wondering it there's an equivalent to echo on somewhere.

Comment (ehsannas) [reply tools/clang/tools/dxcompiler/dxillib.cpp]: I'm sure there's a better way to do it than this, but: temporarily, you can change the appveyor.yml file to run `utils\hct\hcttest -rel` at the `build_script` stage to see the test outputs.

Comment (pow2clk) [reply tools/clang/tools/dxcompiler/dxillib.cpp]: After looking at it closer, I don't think it's the hcttest run that's failing because the powershell script appveyor_test.ps1 prints the last notices and then returns the result at the end. I think the direct invocation of te is what's failing. I could try to swap that in though. Barring that, I can try tweaking the change in a sort of binary search removing changes to try to get a  better handle on what's going on, but I hope it doesn't come to bouncing off appveyor like that.

**Comment by pow2clk** [comment]
I dropped the powershell script in favor of the direct call to TE. It failed in a way that suggests that the clang-hlsl-tests.dll  is missing. It's also possible that the relative path location is wrong. From the output, I think I can anticipate the full path to make sure. Before I do that, I'm going to check to see if there might be any explanation for why that DLL might not be built.

**Comment by pow2clk** [comment]
After some sever abuse of appveyor, I've determined that the reason for the discrepancy between my own system and the appveyor system is the presence of dxil.dll. Something about how it is loaded is causing it to be found invalid. I have dxil.dll on my own system, but I can't seem to get it found by the te execution in question. I'm going to revert all my hacks save one and run one more wasted appveyor run to confirm my conclusion.

File: tools/clang/tools/dxcompiler/dxcapi.cpp [hunk 2]
```diff
@@ -78,11 +78,7 @@ static HRESULT ThreadMallocDxcCreateInstance(
     hr = CreateDxcLibrary(riid, ppv);
   }
   else if (IsEqualCLSID(rclsid, CLSID_DxcValidator)) {
-    if (DxilLibIsEnabled()) {
-      hr = DxilLibCreateInstance(rclsid, riid, (IUnknown**)ppv);
-    } else {
-      hr = CreateDxcValidator(riid, ppv);
-    }
+    hr = CreateDxcValidator(riid, ppv);
```
Comment (pow2clk) [inline tools/clang/tools/dxcompiler/dxcapi.cpp]: I don't consider this an acceptable change. Just part of the testing.

**Comment by pow2clk** [comment]
Okay! This time for sure!

File: include/dxc/Support/dxcapi.use.h [hunk 3]
```diff
@@ -13,38 +13,27 @@
...
-  #endif
-
   HRESULT InitializeInternal(LPCWSTR dllName, LPCSTR fnName) {
-    if (m_dll != nullptr) return S_OK;
-    m_dll = LoadLibraryW(dllName);
-    if (m_dll == nullptr) return HRESULT_FROM_WIN32(GetLastError());
-    m_createFn = (DxcCreateInstanceProc)GetProcAddress(m_dll, fnName);
+    if (m_dll.isValid()) return S_OK;
+    m_dll = DynamicLibrary::getPermanentLibrary(CW2A(dllName));
+    if (!m_dll.isValid()) return E_FAIL;
```
[the remark this answers is not in the document]
Comment (pow2clk) [reply include/dxc/Support/dxcapi.use.h]: This was the cause of the automated testing failure. Getpermanentlibrary makes calls that reset the windows error so this returns S_OK even when the library fails to load.

[the remark this answers is not in the document]
Comment (pow2clk) [reply include/dxc/Support/dxcapi.use.h]: This was the cause of the automated testing failure. Getpermanentlibrary makes calls that reset the windows error so this returns S_OK even when the library fails to load.

[the remark this answers is not in the document]
Comment (ehsannas) [reply include/dxc/Support/dxcapi.use.h]: In that case, I think instead of returning E_FAIL, maybe we should return something more meaningful; closer to the original implementation.
e.g.
```cpp
return HRESULT_FROM_WIN32(ERROR_DLL_INIT_FAILED);
```
or
```cpp
return HRESULT_FROM_WIN32(ERROR_DLL_NOT_FOUND);
```
or
...
This reference has error codes related to DLLs: 
https://msdn.microsoft.com/en-us/library/windows/desktop/ms681383(v=vs.85).aspx

I suppose ERROR_DLL_INIT_FAILED should be okay. (we're not able to check for dll existence here i suppose).

[the remark this answers is not in the document]
Comment (ehsannas) [reply include/dxc/Support/dxcapi.use.h]: What do you think?

File: include/dxc/Support/WinAdapter.h [hunk 4]
```diff
@@ -105,6 +105,9 @@
 #define ERROR_IO_DEVICE EIO
 #define ERROR_INVALID_HANDLE EBADF
 
+// Error returned explicitly. Windows value used.
+#define ERROR_DLL_INIT_FAILED 0x45A
```
Comment (ehsannas) [inline include/dxc/Support/WinAdapter.h]: Since we use std::strerror for displaying errors on Linux, we should use a [valid errno](http://www-numi.fnal.gov/offline_software/srt_public_context/WebDocs/Errors/unix_system_errors.html)
https://github.com/google/DirectXShaderCompiler/blob/bd6553a779750dd42b2e681440809cf16f7d0b09/lib/DxcSupport/dxcapi.use.cpp#L42-L44

I suggest using either ELIBACC or ELIBBAD

Comment (ehsannas) [reply include/dxc/Support/WinAdapter.h]: Well... I guess if we want to return a valid errno, we can't pass it to HRESULT_FROM_WIN32...

Comment (ehsannas) [reply include/dxc/Support/WinAdapter.h]: ughh... since the return type is HRESULT, perhaps we can't use an errno after all...
Should we at least call `SetLastError(ELIBACC)` ?

Comment (ehsannas) [reply include/dxc/Support/WinAdapter.h]: What I would do is manually remove the .so file and run dxc and see what the behavior is...
It should not crash, and ideally it should provide some sort of meaningful message.

Comment (ehsannas) [reply include/dxc/Support/WinAdapter.h]: I can't remember off the top of my head what the stack looks like when handling this error. Depending on how the return value is handled, either returning the errno or HRESULT_FROM_WIN32(errno) or SetLastError(errno) should do it.

**Comment by pow2clk** [comment]
I seem to have missed some of the later comments, but as a result, I learned that MacOS doesn't have ELIBACC. I'll see what I can do.

**Comment by pow2clk** [comment]
Well here's my pitch:

Even just returning `E_FAIL`, I still get a helpful error message:

>error while loading shared libraries: libdxcompiler.so.3.7: cannot open shared object file: No such file or directory

So I'll return `HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND)` since it's close to what we want and present in all systems. The reason there are few library errors and none that are shared by linux and macos is because `dlopen()` failures report errors through `dlerror()` instead of `errno`.
